group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -190,6 +216,118 @@ boolean canTransitionFrom(CoordinatorState state) {
+
+        /**
+         * The topic partition.
+         */
+        final TopicPartition tp;
+
+        /**
+         * The scheduled timers keyed by their key.
+         */
+        final Map<String, TimerTask> tasks = new HashMap<>();
+
+        InternalCoordinatorTimer(TopicPartition tp, LogContext logContext) {
+            this.tp = tp;
+            this.log = logContext.logger(InternalCoordinatorTimer.class);
+        }
+
+        @Override
+        public void schedule(
+            String key,
+            long delay,
+            TimeUnit unit,
+            TimeoutOperation<U> operation
+        ) {
+            TimerTask task = new TimerTask(unit.toMillis(delay)) {
+                @Override
+                public void run() {
+                    String eventName = "Timeout(tp=" + tp + ", key=" + key + ")";
+                    CoordinatorWriteEvent<Void> event = new CoordinatorWriteEvent<>(eventName, tp, coordinator -> {
+                        log.debug("Executing write event {} for timer {}.", eventName, key);
+
+                        // If the task is different, it means that the timer has been
+                        // cancelled while the event was waiting to be processed.
+                        if (!tasks.remove(key, this)) {
+                            throw new RejectedExecutionException("Timer " + key + " was overridden or cancelled");
+                        }
+
+                        // Execute the real operation.
+                        return new CoordinatorResult<>(operation.generateRecords(), null);
+                    });
+
+                    event.future.exceptionally(ex -> {
+                        // Otherwise, we handle the error if the timer is still active.
+                        if (ex instanceof RejectedExecutionException) {
+                            log.debug("The delayed write event {} for the timer {} was not executed because it was " +
+                                "cancelled or overridden.", event.name, key);
+                            return null;
+                        }
+
+                        if (ex instanceof NotCoordinatorException || ex instanceof CoordinatorLoadInProgressException) {
+                            log.debug("The delayed write event {} for the timer {} failed due to {}. Ignoring it because " +
+                                "the coordinator is not active.", event.name, key, ex.getMessage());
+                            return null;
+                        }
+
+                        log.info("The delayed write event {} for the timer {} failed due to {}. Rescheduling it. ",
+                            event.name, key, ex.getMessage(), ex);
+                        schedule(key, 500, TimeUnit.MILLISECONDS, operation);
+
+                        return null;
+                    });
+
+                    log.debug("Scheduling write event {} for timer {}.", event.name, key);
+                    try {
+                        enqueue(event);
```

Comment:
I was having a bit of trouble tracing the code here, but in the exceptional cases where we return null, does the accumulator properly handle this and write no records?

Reply:
Also is this the code that you recently allowed to be null?

Reply:
> I was having a bit of trouble tracing the code here, but in the exceptional cases where we return null, does the accumulator properly handle this and write no records?

I don't follow your comment. What do you mean by "we return null"? Where? Here we enqueue a `CoordinatorWriteEvent` event. If that event returns a CoordinatorResult with records and a null response, the records are written and the event's future is completed with null.

Or are you wondering about the case where the write operation `CoordinatorWriteOperation` would return null? This would result as a NPE in the event handling and the future would be failed with it.

> Also is this the code that you recently allowed to be null?

I have changed CoordinatorResult to accept a null response.

Reply:
In the code above this event.future.exceptionally we return null for most code paths. I was getting confused on how we can return null for that future but still enqueue the event.

Reply:
But I suppose you are saying that null future = null response. 
I guess I wasn't sure how we write records but have a null response and what that means.

Reply:
Are you referring to the nulls returns in exceptionally above? If yes, those are not used anywhere because we don’t use the resulting future.

So we enqueue the event and we react it its result. If it failed, we retry it under some conditions.

Reply:
Hmm -- can you point me to the code that "reacts to its result" and how we don't use the resulting future. (Maybe I'm just confused) but it looks like we always enqueue the event regardless of the future and I don't see where we check if its null and not use it.

Reply:
I was referring to L274. event.future is the future which is completed when (or after) the event is executed (and committed).

When you call event.future.exceptionally, it registers a callback/function that is executed when the future is completed with an exception. Exceptionally returns a new future that is completed with the result of this callback. The returned future is not used here so the null we discussed is not going anywhere.

Reply:
Right, so maybe I need to figure out where this future code is looked at. Because it seems we put this code with the future that could be null in the queue. I guess I'm just struggling to understand why something that was required to be non-null before can just be changed to accept nulls and where that code is.

Reply:
The confusion may come from [here](https://github.com/apache/kafka/pull/13961/files#diff-4b9dd21d7748bfb375634ef6e39037e99f53ed2546519bec32fa1707bd262fe8R271). CoordinatorResult is the part which could accept a null response now. In this particular case, it will be null because we have no response, only records. The null response will be eventually used to complete the event.future. As we use event.future.exceptionally, the null response is never used here either.